src/Microsoft.Azure.ServiceBus/Message.cs

```diff
@@ -6,1136 +6,200 @@ namespace Microsoft.Azure.ServiceBus
-            this.sessionId = sessionId;
-            if (sessionId == null)
-            {
-                this.ClearInitializedMember(MessageMembers.SessionId);
-            }
-            else
-            {
-                this.initializedMembers |= MessageMembers.SessionId;
-            }
-        }
-
-        internal void CopyPartitionKey(string partitionKey)
-        {
-            ValidatePartitionKey("PartitionKey", partitionKey);
-            this.partitionKey = partitionKey;
-            if (partitionKey == null)
-            {
-                this.ClearInitializedMember(MessageMembers.PartitionKey);
-            }
-            else
-            {
-                this.initializedMembers |= MessageMembers.PartitionKey;
-            }
-        }
-
-        internal bool IsMembersSet(MessageMembers members)
-        {
-            bool membersSet = ((this.InitializedMembers & members) != 0);
-
-            return membersSet;
-        }
-
-        static Stream CloneStream(Stream originalStream, bool canThrowException = false)
-        {
-            Stream clonedStream = null;
+            var clone = (Message)this.MemberwiseClone();
+            clone.SystemProperties = new SystemPropertiesCollection();
 
-            if (originalStream != null)
+            if (this.Body != null)
             {
-                MemoryStream memoryStream;
-                ICloneable cloneable;
-
-                if ((memoryStream = originalStream as MemoryStream) != null)
-                {
-                    // Note: memoryStream.GetBuffer() doesn't work
-                    clonedStream = new MemoryStream(memoryStream.ToArray(), 0, (int)memoryStream.Length, false, true);
-                }
-                else if ((cloneable = originalStream as ICloneable) != null)
-                {
-                    clonedStream = (Stream)cloneable.Clone();
-                }
-                else if (canThrowException)
-                {
-                    // TODO: throw Fx.Exception.AsError(new InvalidOperationException(SRClient.BrokeredMessageStreamNotCloneable(originalStream.GetType().FullName)));
-                    throw new InvalidOperationException("BrokeredMessageStreamNotCloneable");
-                }
+                var clonedBody = new byte[this.Body.Count];
+                Array.Copy(this.Body.Array, clonedBody, this.Body.Count);
+                clone.Body = new ArraySegment<byte>(clonedBody);
             }
-
-            return clonedStream;
+            return clone;
         }
 
         /// <summary> Validate message identifier. </summary>
         /// <exception cref="ArgumentException">
         /// Thrown when messageId is null, or empty or greater than the maximum message length.
         /// </exception>
         /// <param name="messageId"> Identifier for the message. </param>
-        static void ValidateMessageId(string messageId)
+        private void ValidateMessageId(string messageId)
```

Comment:
For performance. A single static method as it doesn't need to access any non-static members.